### PR TITLE
Producing dot files

### PR DESCRIPTION
**Pull request by polmiro**
Implementation for #15, adds a new mix task `boundary.visualize` that outputs graphviz diagrams per the spec described in the issue.

Right now it just generates a dummy simple graph output.

### CODE REVIEW

**Comment by polmiro** [comment]
@sasa1977 took a first stab at it, I have some output questions:
* The files are currently written under a new folder called `dot` nested under the root project folder. I assume this may be something that should be configurable. But also maybe something that can be done incrementally. Thoughts?
* The task is writing one schema per parent module found in list of boundaries. I am still not 100% sure if this is the right approach or collect the independent list of graphs to be rendered.

It might be worth it making a more complex test case or adding a few more. Do you happen to have any around that could be handy?

Looking forward to hearing your feedback.

File: lib/mix/tasks/boundary/visualize.ex [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    |> Enum.flat_map(fn %{name: name, deps: deps} ->
+      [name | Enum.map(deps, &elem(&1, 0))]
+    end)
+    |> Enum.uniq()
+    |> Enum.map(&build_node(main_boundary, &1))
+  end
+
+  defp build_node(nil, module), do: {module, :sibling}
+
+  defp build_node(main_boundary, module) do
```
Comment (polmiro) [inline lib/mix/tasks/boundary/visualize.ex:51]: I'm not sure what is the best way to know if the module should be considered one of the siblings in the boundary or not.

File: lib/mix/tasks/boundary/visualize.ex [hunk 2]
```diff
@@ -0,0 +1,51 @@
+defmodule Mix.Tasks.Boundary.Visualize do
+  @shortdoc "Generates a graphviz a dot file for each non-empty boundary."
+  @moduledoc "Generates a graphviz a dot file for each non-empty boundary."
+
+  use Boundary, classify_to: Boundary.Mix
+  use Mix.Task
+
+  @output_folder "dot"
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:8]: I propose we use a bit more specific folder name, like e.g. `boundary`.

File: lib/mix/tasks/boundary/visualize.ex [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      title = format_title(main_boundary)
+      graph = format_graph(title, nodes, edges)
+      file_path = format_file_path(main_boundary)
+
+      File.write!(file_path, graph)
+
+      # image_file_path = "#{Path.rootname(file_path)}.png"
+      # System.cmd("dot", ["-Tpng", file_path, "-o", image_file_path])
+      # System.cmd("open", [image_file_path])
+      # Process.sleep(1000)
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:34]: It's a bit unfortunate this exists in the code. I presume you need it for local testing, right? Couldn't you instead run locally `mix boundary.visualize && pushd dot && dot ... && gnome-open ... && popd`?

File: lib/mix/tasks/boundary/visualize.ex (hunk 1 above)
Comment (sasa1977) [reply lib/mix/tasks/boundary/visualize.ex:51]: A boundary is a sibling if it belongs to the same app and has the same parent. So basically if you invoke `Boundary.fetch!(view, module)` you should get `Boundary.t` as the result, and then you check if the `:app` field is the same, and then that `Boundary.parent` returns the same module for both boundaries.

File: lib/mix/tasks/boundary/visualize.ex [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    end
+  end
+
+  defp build_edges(_main_boundary, boundaries) do
+    boundaries
+    |> Enum.flat_map(fn %{name: name, deps: deps} ->
+      Enum.map(deps, fn {dep_name, mode} ->
+        {name, dep_name, mode}
+      end)
+    end)
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:65]:
```elixir
for boundary <- boundaries, {dep_name, mode} <- deps, do: {boundary.name, dep_name, mode}
```

File: lib/mix/tasks/boundary/visualize.ex [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  defp build_node(main_boundary, module) do
+    if Module.split(main_boundary.name) == Module.split(module) |> Enum.drop(-1) do
+      {module, :sibling}
+    else
+      {module, nil}
+    end
+  end
+
+  defp build_edges(_main_boundary, boundaries) do
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:59]: Why is this function taking main_boundary when it's not used?

File: lib/mix/tasks/boundary/visualize.ex [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  defp format_nodes(nodes) do
+    nodes
+    |> Enum.sort()
+    |> Enum.map(&format_node_description/1)
+    |> Enum.join(";\n  ")
+  end
+
+  def format_node_description({module, :sibling}), do: format_node(module)
+  def format_node_description({module, _}), do: "#{format_node(module)} [color = \"gray\"]"
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:97]: You could use `~s// ` and then you don't need to escape inner quotes.

File: lib/mix/tasks/boundary/visualize.ex [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+  defp format_edge(edge = {module1, module2, _}) do
+    "#{format_node(module1)} -> #{format_node(module2)}#{format_edge_attributes(edge)}"
+  end
+
+  def format_node(module) when is_atom(module) do
+    "\"#{format_name(module)}\""
+  end
+
+  defp format_name(module) when is_atom(module) do
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:114]: This guard is redundant since the function always takes an atom. The same holds for `format_node`.

File: lib/mix/tasks/boundary/visualize.ex [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    module
+    |> Module.split()
+    |> Enum.join(".")
```
Comment (sasa1977) [inline lib/mix/tasks/boundary/visualize.ex:117]: Same as `inspect(module)`.

File: test/mix/tasks/visualize_test.exs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+          labelloc=top;
+
+          "BlogEngine";
+          "BlogEngineApp";
+          "BlogEngineWeb";
+          "Mix";
+
+          "BlogEngineApp" -> "BlogEngine";
+          "BlogEngineApp" -> "BlogEngineWeb";
+          "BlogEngineApp" -> "Mix" [label = \"compile\"];
```
Comment (sasa1977) [inline test/mix/tasks/visualize_test.exs:51]: No need to escape double quotes inside a heredoc.

File: test/mix/tasks/visualize_test.exs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+          label="BlogEngine boundary";
+          labelloc=top;
+
+          "BlogEngine.Accounts";
+          "BlogEngine.Articles";
+          "BlogEngine.Repo";
+          "Mix" [color = \"gray\"];
+
+          "BlogEngine.Accounts" -> "BlogEngine.Repo";
+          "BlogEngine.Accounts" -> "Mix" [label = \"compile\"];
```
Comment (sasa1977) [inline test/mix/tasks/visualize_test.exs:70]: I'm thinking that it would be nice draw an edge to a non-sibling boundary as dashed and gray. What do you think?

File: lib/mix/tasks/boundary/visualize.ex (hunk 3 above)
Comment (polmiro) [reply lib/mix/tasks/boundary/visualize.ex:34]: Definitely, I just forgot I had that there when I was playing around with the visualizations. Thanks!

File: lib/mix/tasks/boundary/visualize.ex (hunk 4 above)
Comment (polmiro) [reply lib/mix/tasks/boundary/visualize.ex:65]: much nicer, I must start getting more used to using list comprehensions 👍

**Comment by sasa1977** [comment]
> I think this is now good enough for the initial version. I'll merge this as it is, and do some final polish. Thank you for contributing!

Well, I see now that you marked this PR as a WIP, so before merging, did you have anything else you wanted to implement here?

**Comment by polmiro** [comment]
Great, that's all. Cheers!
